Added support for o1-2024-12-17 and structured outputs

shahofblah opened the pull request:
1. Added o1-2024-12-17 to the list of models(the older models in the list don't support structured outputs) 
2. Added code to rename max_tokens to max_completion tokens if model belongs to o1 family
3. Added support for [OpenAI Structured Outputs](https://platform.openai.com/docs/guides/structured-outputs#supported-models)
    a) Added response_format in LLMParams(currently ignored for caching/serialisation)
    b) Switched to OpenAI parse instead of completion API if response_format is a subclass of pydantic.BaseModel(OpenAI complains if we send a classtype response_format in the completion API)

dpaleka commented on safetytooling/apis/inference/openai/utils.py:
```diff
@@ -29,6 +29,8 @@
     "o1-mini-2024-09-12",
     "o1-preview",
     "o1-preview-2024-09-12",
+    # Models older than 2024-12-17 do not support structured outputs
```
Doesn't seem to be true? https://platform.openai.com/docs/guides/structured-outputs#supported-models

While you're there, add all recent models

shahofblah replied:
I'd meant within the o1 family. Changed to clarify

I think we already have the [latest models](https://platform.openai.com/docs/models#current-model-aliases)

dpaleka commented on safetytooling/data_models/inference.py:
```diff
@@ -17,13 +17,21 @@ class LLMParams(HashableBaseModel):
     num_candidates_per_completion: int = 1
     insufficient_valids_behaviour: Literal["error", "continue", "pad_invalids", "retry"] = "retry"
-    max_tokens: int | None = None
+    max_tokens: int | None = None  # For o1 models this is converted to max_completion_tokens in the API layer
```
In OpenAI API this is deprecated in favor of max_completion_tokens. Why not just always rename it for OpenAI models?
https://platform.openai.com/docs/api-reference/chat/create#chat-create-max_completion_tokens

You can also have the two params and assert only one is present.

shahofblah replied:
Done
Didn't add 2 params as wanted to avoid caller-side complexity

dpaleka commented on safetytooling/data_models/inference.py:
```diff
@@ -17,13 +17,21 @@ class LLMParams(HashableBaseModel):
     top_k_logprobs: int | None = None  # ?
     logit_bias: dict[int, float] | None = None
-
+    response_format: Any = pydantic.Field(
```
Write a test for this with gpt-4o-mini-2024-07-18, see how it's done in tests/

shahofblah replied:
Added test_gpt4o_mini_2024_07_18_with_response_format in test_api.py

dpaleka commented on safetytooling/apis/inference/api.py:
```diff
@@ -406,8 +405,11 @@ async def __call__(
-        else:
-            if max_tokens is not None:
+        elif max_tokens is not None:
+            if any(model_id.startswith("o1") for model_id in model_ids):
```
See my comment in the args config -- this should be renamed for all OpenAI models

shahofblah replied:
Changed

dpaleka commented on safetytooling/apis/inference/api.py:
```diff
@@ -369,8 +369,7 @@ async def __call__(
             prompt: The prompt to send to the model(s). Type should be Prompt.
             audio_out_dir : The directory where any audio output from the model (relevant for speech-to-speech models) should be saved
-            max_tokens: The maximum number of tokens to request from the API (argument added to
-                standardize the Anthropic and OpenAI APIs, which have different names for this).
```
Note that there was some comment here that conflicts with the new comment/functionality.

shahofblah replied:
Reworded for accuracy

shahofblah commented:
Addressed comments
1. Instead of renaming max_tokens to max_completion tokens specifically for the o1 family, I now rename it for all OpenAI models. Changed comments to reflect this too
2. Added a test for structured output in response_format

jplhughes commented on safetytooling/apis/inference/openai/utils.py:
```diff
@@ -29,6 +29,9 @@
     "o1-mini-2024-09-12",
     "o1-preview",
     "o1-preview-2024-09-12",
+    # THese o1 models support structured outputs https://platform.openai.com/docs/guides/structured-outputs#supported-models
+    "o1",
+    "o1-2024-12-17",
```
@shahofblah  @dpaleka there are quite a few other places in this file where new models need to be added (e.g. for costs and rate limits)

dpaleka replied:
Let's see if Devin can do this one
